Document driver testing status

mpg opened the pull request:
## Description

This PR's main goal is to document the strategy used to test the PSA Crypto driver interface, as well as the current implementation status of that strategy.

Along the way, a couple of minor things are modified, but the PR's goal is not to fix things, or even to provide a complete plan for action - only to document the current state of things (which could be the first step towards an action plan).

Main gaps identified:
- Opaque is mostly not tested at all (only a few key management entry points are tested but not even all of them (edit: and [now](https://github.com/Mbed-TLS/mbedtls/pull/8700) asymmetric encrypt/decrypt), and not simulating an opaque driver with internal storage).
- Key derivation is not tested at all.
- Fallback to other entry points for operations that support it (like using mulit-part entry points for one-short API when the driver doesn't provide one-shot) is not tested at all - see also #4570
- ~Two gaps in driver-only testing (compared to what we're supposed to support).~ Fixed in the meantime: [8616](https://github.com/Mbed-TLS/mbedtls/pull/8616), [8734](https://github.com/Mbed-TLS/mbedtls/pull/8734).
- A number of gaps in `test_suite_psa_crypto_driver_wrapper.data`. 

## PR checklist

- [x] **changelog** not required - only documentation and testing
- [x] **backport** not required - only useful for new development
- [x] **tests** not required - only documentation and testing changes

mpg commented:
@gilles-peskine-arm @ronald-cron-arm Since you're probably the most familiar with this, I'd appreciate if you could take a look - maybe not a complete review but at high-level view to see if this looks reasonable or if you notice any misunderstanding or obvious omissions here.

valeriosetti reviewed (commented):
I only found 1 typo, the rest is fine for me. However the CI is not fully happy about the change made to `generate_ec_key`

valeriosetti reviewed (approved):
LGTM

valeriosetti commented:
I think that at least 3 out of 4 errors found in `test_psa_crypto_config_accel_cipher_aead` belong to the fact that CTR-DRBG is using AES through PSA now:
```
[2023-12-18T11:12:54.941Z] generate_ec_key through transparent driver: fake .................. FAILED
[2023-12-18T11:12:54.941Z]   mbedtls_test_driver_key_management_hooks.hits == 1
[2023-12-18T11:12:54.941Z]   at line 791, suites/test_suite_psa_crypto_driver_wrappers.function
[2023-12-18T11:12:54.941Z]   lhs = 0x0000000000000005 = 5
[2023-12-18T11:12:54.941Z]   rhs = 0x0000000000000001 = 1
[2023-12-18T11:12:54.941Z] generate_ec_key through transparent driver: in-driver ............. FAILED
[2023-12-18T11:12:54.941Z]   mbedtls_test_driver_key_management_hooks.hits == 1
[2023-12-18T11:12:54.941Z]   at line 791, suites/test_suite_psa_crypto_driver_wrappers.function
[2023-12-18T11:12:54.941Z]   lhs = 0x0000000000000008 = 8
[2023-12-18T11:12:54.941Z]   rhs = 0x0000000000000001 = 1
[2023-12-18T11:12:54.941Z] generate_ec_key through transparent driver: fallback .............. FAILED
[2023-12-18T11:12:54.941Z]   mbedtls_test_driver_key_management_hooks.hits == 1
[2023-12-18T11:12:54.941Z]   at line 791, suites/test_suite_psa_crypto_driver_wrappers.function
[2023-12-18T11:12:54.941Z]   lhs = 0x0000000000000008 = 8
[2023-12-18T11:12:54.941Z]   rhs = 0x0000000000000001 = 1
```
I think we should fix this (as we did recently) by adding a specific hit counter instead of `mbedtls_test_driver_key_management_hooks.hits`. Wdyt?

mpg commented:
> I think we should fix this (as we did recently) by adding a specific hit counter instead of `mbedtls_test_driver_key_management_hooks.hits`. Wdyt?

Sounds good, thanks for the suggestion! I hadn't even looked at the CI results yet, thanks for analyzing the failures!

valeriosetti reviewed (commented):
I only found 1 typo which I didn't see before. The rest is fine for me :)

mpg commented:
Note to self: update names of cipher/aead components when #8641 has been merged.

valeriosetti commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -128,6 +128,455 @@ Every cryptographic mechanism for which a transparent driver interface exists (k
+The transparent driver fully implements the declared entry points, and can use
+any backend: internal or libtestdriver1.
+
+The opaque driver is not implemented at all, neither instumentation nor the
```
I think this is not correct: I think I implemented opaque encryption/decryption in #8700. Isn't it?

mpg replied:
Ah great, I wasn't aware. I'll update the document then.

Note: this why I'd like to get this document merged soonish. Then the burden of updating it will be on PRs that change things.

gilles-peskine-arm commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -128,6 +128,453 @@ Every cryptographic mechanism for which a transparent driver interface exists (k
+Not covered at all by the test drivers.
+
+That's a gap in our testing, as the driver interface does define a key
+derivation family of entry points. This gap is probably related to the fact
+that our internal code structure doesn't obey the guidelines and is not
+aligned with the driver interface, see #5488 and related issues.
```
Actually, it's because we don't support driver dispatch at all for key derivation. The driver interface specification covers key derivation, but we haven't implemented that part.

mpg replied:
This was my way of saying we don't support driver dispatch. I'll rephrase.

gilles-peskine-arm commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -128,6 +128,453 @@ Every cryptographic mechanism for which a transparent driver interface exists (k
+back to the built-in implementation when the driver returns `NOT_SUPPORTED).
+
+This is `test_suite_psa_crypto_driver_wrappers`, which is maintained manually
+(that is, the test cases in the `.data` files are not auto-generated). The
```
Can you please add a sentence or so about the validation strategy? At a glance, it seems to be done with mostly with known answer tests, and omits randomized algorithms, except for asymmetric encrypt/decrypt which uses encrypt/decrypt consistency, right?

gilles-peskine-arm reviewed (changes requested):
I've done a first pass where I reviewed everything except for the details of which entry points are covered in this or that place. Thanks for writing this up, it's very useful both as an explanation of what we have and as a basis for understanding and filling the test gaps.

I've requested a few minor updates but this looks good overall.

I'm not even sure a complete review of entry point coverage is necessary: now that you've done the boring work (sorry!), we have a good idea of how much is missing and that's good enough to estimate the remaining work even if the details aren't 100% right on a function-by-function basis. Also there's a risk that the information will become obsolete because we tend to forget to update this kind of documentation when we make incremental code improvements.

ronald-cron-arm commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -128,6 +128,453 @@ Every cryptographic mechanism for which a transparent driver interface exists (k
+Historical note: internal was initially the only back-end; then support for
+libtestdriver1 was added gradually.
+
+Question: if/when we have complete libtestdriver1 support, do we still need
```
Yes I think the same, if internal does not bring anything compared to libtestdriver1 in terms of supported entry points we should just remove it.

mpg replied:
Note: I think right now `component_test_psa_crypto_drivers` is not yet redundant with the various `libtestdriver1`-based components, because those are driver-only, while `component_test_psa_crypto_drivers` exercises fallback. But we could easily do that with `libtestdriver1` too. We just need to make sure we don't remove testing for fallback.

(Note: once we finally have `analyze_outcomes.py` enforcement, there will be no risk, as the "fallback" test cases would show as not executed if we no loner had a component running them.)

mpg commented on tests/configs/user-config-for-test.h:
```diff
@@ -37,34 +37,76 @@
 /* Use the accelerator driver for all cryptographic mechanisms for which
- * the test driver implemented. */
+ * the test driver is implemented. This is copied from psa/crypto_config.h
+ * with the parts not implemented by the test driver commented out. */
```
I agree with the general point, but do you really mean PAKE? It seems to me it's just KDFs and "cooked" key derivation.

Anyway, created https://github.com/Mbed-TLS/mbedtls/issues/9027

mpg commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -128,6 +128,453 @@ Every cryptographic mechanism for which a transparent driver interface exists (k
+
+#### Test case coverage
+
+Since `test_suite_psa_crypto_driver_wrappers.data` is maintained manually,
```
Agreed, that's what I was hinting at with "since it's maintained manually, how can we be sure?". But explicit is better than implicit, so I'll expand.

mpg commented:
@gilles-peskine-arm Thanks for your review! I believe I have addressed your feedback.

I've also had to rebase to resolve a conflict. Turns out the commit "Fix failure in some configs by being more precise" was no longer needed as something similar had been done in development in the meantime.

gilles-peskine-arm commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -181,8 +177,10 @@ they're available);
+implementing the operation. This will remain hard to test until we have proper
+support for JSON-defined drivers with auto-generation of dispatch code.
+(The `MBEDTLS_PSA_ACCEL_xxx` macros we currently use are not expressive enough
+to specify which entry points are support for a given mechanism.)
```
```suggestion
to specify which entry points are supported for a given mechanism.)
```

gilles-peskine-arm commented on docs/architecture/testing/driver-interface-test-strategy.md:
```diff
@@ -384,8 +384,10 @@ The following entry points are declared (transparent and opaque):
-operation: entry points always return `NOT_SUPPORTED`.
+The opaque driver implements the declared entry points, and can use any
+backend: internal or libtestdriver1. However it does not implement the
+instrumentation (hits, forced output/status), as this [was not an immediate
```
From a chat with Ronald: hits is actually probably an obsolete concern. With the internal backend, hits were how we knew that the driver was called, as opposed to directly calling the built-in code. With libtestdriver1, we check that by ensuring that the built-in code is not present, so if the operation gives the correct result, only a driver call can have calculated that result. So there is low value in checking the hit count. There is still some value for hit counts, e.g. checking that we don't call a multipart entry point when we intended to call the one-shot entry point, but it's limited.

gilles-peskine-arm reviewed (changes requested):
LGTM except for a typo, and maybe one bit of explanation to add.

gilles-peskine-arm reviewed (approved):
LGTM

valeriosetti reviewed (approved):
LGTM ;)
